src/cbmc/bmc_util.cpp

```diff
@@ -0,0 +1,267 @@
+    graphml_witness(symex_target_equation);
+  else
+    return;
+
+  if(graphml == "-")
+    write_graphml(graphml_witness.graph(), std::cout);
+  else
+  {
+    std::ofstream out(graphml);
+    write_graphml(graphml_witness.graph(), out);
+  }
+}
+
+void convert_symex_target_equation(
+  symex_target_equationt &equation,
+  prop_convt &prop_conv,
+  message_handlert &message_handler)
+{
+  messaget msg(message_handler);
+  msg.status() << "converting SSA" << messaget::eom;
+
+  // convert SSA
+  equation.convert(prop_conv);
+}
+
+std::unique_ptr<memory_model_baset>
+get_memory_model(const optionst &options, const namespacet &ns)
+{
+  const std::string mm = options.get_option("mm");
+
+  if(mm.empty() || mm == "sc")
+    return util_make_unique<memory_model_sct>(ns);
+  else if(mm == "tso")
+    return util_make_unique<memory_model_tsot>(ns);
+  else if(mm == "pso")
+    return util_make_unique<memory_model_psot>(ns);
+  else
+  {
+    throw "invalid memory model '" + mm + "': use one of sc, tso, pso";
+  }
+}
+
+void setup_symex(
+  symex_bmct &symex,
+  const namespacet &ns,
+  const optionst &options,
+  ui_message_handlert &ui_message_handler)
+{
+  messaget msg(ui_message_handler);
+  const symbolt *init_symbol;
+  if(!ns.lookup(INITIALIZE_FUNCTION, init_symbol))
+    symex.language_mode = init_symbol->mode;
+
+  msg.status() << "Starting Bounded Model Checking" << messaget::eom;
+
+  symex.last_source_location.make_nil();
+
+  symex.unwindset.parse_unwind(options.get_option("unwind"));
+  symex.unwindset.parse_unwindset(options.get_option("unwindset"));
+}
+
+void slice(
+  symex_bmct &symex,
+  symex_target_equationt &symex_target_equation,
+  const namespacet &ns,
+  const optionst &options,
+  ui_message_handlert &ui_message_handler)
+{
+  messaget msg(ui_message_handler);
+  if(options.get_option("slice-by-trace") != "")
+  {
+    symex_slice_by_tracet symex_slice_by_trace(ns);
+
+    symex_slice_by_trace.slice_by_trace(
+      options.get_option("slice-by-trace"), symex_target_equation);
+  }
+  // any properties to check at all?
+  if(symex_target_equation.has_threads())
+  {
+    // we should build a thread-aware SSA slicer
+    msg.statistics() << "no slicing due to threads" << messaget::eom;
+  }
+  else
+  {
+    if(options.get_bool_option("slice-formula"))
+    {
+      ::slice(symex_target_equation);
+      msg.statistics() << "slicing removed "
+                       << symex_target_equation.count_ignored_SSA_steps()
+                       << " assignments" << messaget::eom;
+    }
+    else
+    {
+      if(options.get_list_option("cover").empty())
+      {
+        simple_slice(symex_target_equation);
+        msg.statistics() << "simple slicing removed "
+                         << symex_target_equation.count_ignored_SSA_steps()
+                         << " assignments" << messaget::eom;
+      }
+    }
+  }
+  msg.statistics() << "Generated " << symex.get_total_vccs() << " VCC(s), "
```

Comment:
It looks like `symex_target_equation` is assumed to be bound to `symex` at this point, such that slicing the target equation has an invisible side-effect on `symex` (so that `get_remaining_vccs()` reflects the result of slicing). That suggests moving this onto `goto_symext`